### ISSUE TITLE
Host system headers found/used when cross-compiler specified

### ISSUE THREAD

**Issue by UberUser** [microsoft/vscode-cpptools#5797]
Issue Type: <b>Bug</b>

Hi,

First of all, I have to say that Code is amazing! IDEs that will support non-native toolchains on Linux with remote display on a Mac (however accomplished) are few. All I've tried compromised on some aspect of the full experience. Until Code. It's the fastest, best looking, least piggy, and least hassle. I'd almost pay if I had to. :)

There is however one small thing keeping it from perfection. I use a MIPS cross-compiler hosted on RHEL 7.6. In the C/C++ extension I've set the compiler path and a few compiler arguments (not important, just -mabi=32 -EL).

This works in that it then finds some headers in my cross-compiler sysroot. However, it still continues to find host headers in /usr/include.

E.g. with fs/inode.c open (while browsing a full Linux kernel) and selecting Go to Definition on the first include file (<linux/fs.h>) what gets opened is /usr/include/linux/fs.h. Similarly for <linux/time.h>.

The same problem will show up indirectly: I get problems (shown in the PROBLEMS pane) such as

cannot open source file "gnu/stubs-32.h" (dependency of "linux/pagemap.h")

There is no gnu/stubs-32.h in my sysroot, but there is one in /usr/include. Opening the half dozen nested includes along the way from linxu/pagemap.h to gnu/stubs-32.h reveals a jump from my sysroot to /usr/include along the way.

This should be easy to solve: I think that if a compiler with its own sysroot is set, then no host includes should ever be looked at. If for some situations that wouldn't be correct, then a checkbox next to the compiler path along the lines of "only use headers from this compiler's sysroot" or "ignore host includes" would completely solve the problem.

Thanks for a great product!

Extension version: 0.29.0
VS Code version: Code 1.47.0 (d5e9aa0227e057a60c82568bf31c04730dc15dcd, 2020-07-09T08:01:54.115Z)
OS version: Darwin x64 18.7.0
Remote OS version: Linux x64 3.10.0-957.21.3.el7.x86_64

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz (8 x 2800)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>viz_display_compositor: enabled_on<br>viz_hit_test_surface_layer: disabled_off_ok<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|4, 4, 3|
|Memory (System)|16.00GB (1.07GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|

|Item|Value|
|---|---|
|Remote|SSH: redacted|
|OS|Linux x64 3.10.0-957.21.3.el7.x86_64|
|CPUs|Intel(R) Xeon(R) CPU E5-2609 v2 @ 2.50GHz (7 x 1782)|
|Memory (System)|125.55GB (58.95GB free)|
|VM|0%|
</details>

**Comment by UberUser** [comment]
"There is no gnu/stubs-32.h in my sysroot, but there is one in /usr/include."

Whoops, I misstated this. There is no /usr/include/gnu/stubs-32.h installed on the host used in the reported configuration either.

What's germane is that there is no reference to "gnu/stubs-32.h" anywhere in my sysroot headers (but it is included from /usr/include/gnu/stubs.h). It's only looked for because somewhere along the line while following nested includes a jump is made from finding headers in the compiler sysroot to finding them in /usr/include.

The 32 bit stubs.h is looked for because I set the Intellisense mode to gnu-x86, due to the lack of a mips32 or a generic 32 bit mode. (That's another wish I was going to save for another issue.)

**Comment by sean-mcmanus** [comment]
Sounds like https://github.com/microsoft/vscode-cpptools/issues/5035 ?

**Comment by UberUser** [comment]
> Sounds like #5035 ?

Why yes, that request for MIPS support covers my 2nd little add-on request. (I assume if any MIPS support were added it would be for both 32 and 64 bit.)

**Comment by sean-mcmanus** [comment]
Okay, I'm a little confused on how I can repro this or diagnose what is going unexpected. Can you run C/C++: Log Diagnostics and see if the Includes has the sysroot paths or not?

Oh, that other request was for MIPS64, but if we ever get around to supporting MIPS we might add a new issue for tracking 32-bit as well. We're currently working on ARM support (MIPS hasn't gotten enough requests yet).

**Comment by UberUser** [comment]
> Okay, I'm a little confused on how I can repro this or diagnose what is going unexpected. Can you run C/C++: Log Diagnostics and see if the Includes has the sysroot paths or not?

I'll bet you could repo it by setting "Compiler path" to any cross compiler (or perhaps it's only gcc), opening a trivial program such as "hello world", and observing that <stdio.h> will resolve to /usr/include/stdio.h, instead of wherever the cross compiler's sysroot is located (as determined by "gcc -print-sysroot").

The log diagnostics show

Includes:
        /usr/include/c++/4.8.2
        /usr/local/include
        /usr/include

followed by the include directories in the workspace path. And I wasn't paying attention earlier: my sysroot include directories are not listed at all. I was too quickly glancing at paths only looking for whether or not they were /usr/include, and didn't notice all of the other include paths were in Linux.

In my attempts to avoid the system includes, I removed the default "Linux" configuration, leaving only my cross-compiler one. No difference.

**Comment by sean-mcmanus** [comment]
We query the compiler for system includes. What is output when you use "gcc -Wp,-v -E -x c++ /dev/null"? A similar issue might be https://github.com/microsoft/vscode-cpptools/issues/5113 .

**Comment by UberUser** [comment]
And here I thought that it was the headers in the sysroot that were used. I'm happy to find out about this other set.

The requested output plus an example of the paths simplified:

```
$ mipsel-linux-gcc  -Wp,-v -E -x c++ /dev/null
ignoring nonexistent directory "/longpath/mipsel-buildroot-linux-uclibc/sysroot/usr/local/include"
#include "..." search starts here:
#include <...> search starts here:
 /longpath/lib/gcc/mipsel-buildroot-linux-uclibc/7.3.0/../../../../mipsel-buildroot-linux-uclibc/include/c++/7.3.0
 /longpath/lib/gcc/mipsel-buildroot-linux-uclibc/7.3.0/../../../../mipsel-buildroot-linux-uclibc/include/c++/7.3.0/mipsel-buildroot-linux-uclibc
 /longpath/lib/gcc/mipsel-buildroot-linux-uclibc/7.3.0/../../../../mipsel-buildroot-linux-uclibc/include/c++/7.3.0/backward
 /longpath/lib/gcc/mipsel-buildroot-linux-uclibc/7.3.0/include
 /longpath/lib/gcc/mipsel-buildroot-linux-uclibc/7.3.0/include-fixed
 /longpath/lib/gcc/mipsel-buildroot-linux-uclibc/7.3.0/../../../../mipsel-buildroot-linux-uclibc/include
 /longpath/mipsel-buildroot-linux-uclibc/sysroot/usr/include
End of search list.
# 1 "/dev/null"
# 1 "<built-in>"
# 1 "<command-line>"
# 1 "/dev/null"

$ pushd /longpath/lib/gcc/mipsel-buildroot-linux-uclibc/7.3.0/../../../../mipsel-buildroot-linux-uclibc/include/c++/7.3.0
$ pwd
/longpath/mipsel-buildroot-linux-uclibc/include/c++/7.3.0
```

**Comment by sean-mcmanus** [comment]
Is your compilerPath set to the mipsel-linux-gcc compiler? From the Includes logging earlier it appears it's not getting used. If you set C_Cpp.loggingLevel to "Debug" do you see any more compiler querying diagnostics in the C/C++ logging that can explain why it appears to be falling back to some other compiler with gcc 4.8 headers?

**Comment by UberUser** [comment]
> Is your compilerPath set to the mipsel-linux-gcc compiler?

Yes

> From the Includes logging earlier it appears it's not getting used. If you set C_Cpp.loggingLevel to "Debug" do you see any more compiler querying diagnostics in the C/C++ logging that can explain why it appears to be falling back to some other compiler with gcc 4.8 headers?

I assumed that setting goes in .vscode/c_cpp_properties.json. I added
`"C_Cpp.loggingLevel": "Debug"`
to the end of my one configuration, and ran Log: Diagnostics again.

The only change in output was in the process info and the addition of the logging level setting itself.  That makes me wonder if I did that right...

Further details: I temporarily replaced the compiler path with a script that just records its arguments. It gets executed with the options you had me use above (plus the one for macros).

Is there any chance that the fact that the compiler's query output begins with `ignoring nonexistent directory` is causing the rest of it to be ignored?

Finally I tried changing the mode to gcc-arm from gcc-x86 to avoid another potential excuse to be looking at my x86 host system headers. No difference there either.

**Comment by sean-mcmanus** [comment]
By "C/C++ logging" I mean the "C/C++" logging section and not the "C/C++ Diagnostics" section.

The "ignoring nonexistent directory" is commonly occurring and unlikely to cause an issue.

**Comment by UberUser** [comment]
> By "C/C++ logging" I mean the "C/C++" logging section and not the "C/C++ Diagnostics" section.

Er, Ok, but where is that? I don't see it as as command to run nor is it a choice in the drop down that includes the "C/C++ Diagnostics" section:

![Screen Shot 2020-07-27 at 5 23 25 PM](https://user-images.githubusercontent.com/68451404/88605386-954ac600-d02e-11ea-8d62-a66d7a1f06e2.png)

With the last choice listed there though there's the following at the end of the output:

```
[2020-07-27 16:46:31.121] [exthost] [info] ExtensionService#loadCommonJSModule file://.../.vscode-server/bin/17299e413d5590b14ab0340ea477cdd86ff13daf/extensions/npm/dist/npmMain
[2020-07-27 17:29:14.583] [exthost] [error] [ms-vscode.cpptools] provider FAILED
[2020-07-27 17:29:14.583] [exthost] [error] undefined
```

That seems like it ought to matter to something, even if not to this headers problem.

**Comment by sean-mcmanus** [comment]
I don't see the "C/C++" logging section in your screenshot. Are you sure the loggingLevel is set to "Debug" and the extension is running? See https://code.visualstudio.com/docs/cpp/enable-logging-cpp .

The [ms-vscode.cpptools] provider FAILED error message is okay and unrelated. There's another VS Code issue tracking that.

**Comment by UberUser** [comment]
> I don't see the "C/C++" logging section in your screenshot. Are you sure the loggingLevel is set to "Debug" and the extension is running?

No, I'm not sure at all. That's why I said "I assumed that setting goes in .vscode/c_cpp_properties.json." and said I added it there. However with the directions from your link I see that that setting goes in .vscode/settings.json. With it set there I get

```
File exclude: **/.git
File exclude: **/.svn
File exclude: **/.hg
File exclude: **/CVS
File exclude: **/.DS_Store
File exclude: **/.vscode
Search exclude: **/node_modules
Search exclude: **/bower_components
Search exclude: **/*.code-search
Search exclude: **/.vscode
Populate include completion cache.
Discovering files...
  Processing folder (recursive): /usr/local/include/
  Processing folder (recursive): /usr/include/
  Processing folder (recursive): /longpath/linux-4.4/
  Discovering files: 28506 file(s) processed
  0 file(s) removed from database
Done discovering files.
Parsing remaining files...
  Parsing: 0 files(s) processed
Done parsing remaining files.
idle loop: reparsing the active document
Checking for syntax errors: file:///longpath/linux-4.4/fs/inode.c
Queueing IntelliSense update for files in translation unit of: /longpath/linux-4.4/fs/inode.c
Error squiggle count: 449
Update IntelliSense time (sec): 0.3

```

**Comment by sean-mcmanus** [comment]
Your logging appears to be missing info. Can you try opening the workspace without any files open, and then open inode.c?

**Comment by UberUser** [comment]
> Your logging appears to be missing info. Can you try opening the workspace without any files open, and then open inode.c?

I quit and restarted, and yea! I get a couple of very clear error messages. Full log after my comments on it.

The "unrecognized argument" is because I had my compiler options all on one line. I happily pasted them in there without noticing the instruction "One argument per line". Doh! So Ok, after correcting that, much happiness, and parsing too. The compiler and sysroot headers were found and scanned, and I see no evidence that the host headers are being looked at.

As for the other error "Compiler does not support 32-bit.", that's not true. It's only useable for 32-bit, as there are only 32-bit versions of the system libraries and linker objects available. This error remains after fixing my compiler arguments. I can open a new issue for that if you don't have a stock answer at the ready. In the meantime, is there a setting for using 32-bit regardless of what the compiler says or not?

My user level code (for which I'm using a different Code configuration) still has the original problem even after I  fixed the compiler arguments problem in it. I will go back and set that up again from scratch, and if I can't figure it out with all of the debugging info I now know about I'll open a new issue.

Many thanks!

P.S. I would not mind at all if there was less falling back going on here... if one provides a compiler and the info for it cannot be gotten, saying so and giving up would let one know there was as problem instead of completely hiding it. It would save time in the long run too, by avoiding going through a debugging exchange such as this.

Logging that showed the error:

cpptools/didChangeCppProperties
Attempting to get defaults from compiler in "compilerPath" property: '/toolchainpath/bin/mipsel-linux-gcc'
Compiler does not support 32-bit. Falling back to 64-bit intelliSenseMode.
terminating child process: 593
Failed to query compiler. Falling back to no bitness.
terminating child process: 597
mipsel-linux-gcc.br_real: error: unrecognized argument in option ‘-mabi=32 -EL’
mipsel-linux-gcc.br_real: note: valid arguments to ‘-mabi=’ are: 32 64 eabi n32 o64

terminating child process: 598
Code browsing service initialized
Attempting to get defaults from compiler in "compilerPath" property: '/toolchainpath/bin/mipsel-linux-gcc'
Compiler does not support 32-bit. Falling back to 64-bit intelliSenseMode.
terminating child process: 599
Failed to query compiler. Falling back to no bitness.
terminating child process: 600
mipsel-linux-gcc.br_real: error: unrecognized argument in option ‘-mabi=32 -mips32r2 -mtune=24kc -EL’
mipsel-linux-gcc.br_real: note: valid arguments to ‘-mabi=’ are: 32 64 eabi n32 o64

terminating child process: 601
  Folder: /usr/local/include/ will be indexed
  Folder: /usr/include/ will be indexed
  Folder: /longpath/linux-4.4/ will be indexed
Populate include completion cache.
Discovering files...
  Processing folder (recursive): /usr/local/include/
  Processing folder (recursive): /usr/include/
  Processing folder (recursive): /longpath/linux-4.4/
  Discovering files: 28506 file(s) processed
  0 file(s) removed from database
Done discovering files.
Parsing remaining files...
  Parsing: 0 files(s) processed
Done parsing remaining files.
cpptools/getCodeActions: file:///longpath/linux-4.4/fs/inode.c (id: 2)
textDocument/didOpen: file:///longpath/linux-4.4/fs/inode.c
cpptools/activeDocumentChange: file:///longpath/linux-4.4/fs/inode.c
cpptools/textEditorSelectionChange
cpptools/getDocumentSymbols: file:///longpath/linux-4.4/fs/inode.c (id: 3)
cpptools/getDocumentSymbols
cpptools/textEditorSelectionChange
cpptools/getDocumentSymbols: file:///longpath/linux-4.4/fs/inode.c (id: 4)
cpptools/getDocumentSymbols: file:///longpath/linux-4.4/fs/inode.c (id: 5)
cpptools/getSemanticTokens: file:///longpath/linux-4.4/fs/inode.c (id: 6)
cpptools/getCodeActions: file:///longpath/linux-4.4/fs/inode.c (id: 7)
cpptools/getFoldingRanges: file:///longpath/linux-4.4/fs/inode.c (id: 8)
sending compilation args for /longpath/linux-4.4/fs/inode.c
  include: /usr/include/c++/4.8.2
  include: /usr/local/include
  include: /usr/include
  include: /longpath/linux-4.4/include
  include: /longpath/linux-4.4/arch/mips/include
  include: /longpath/linux-4.4/arch/mips/include/asm/mach-ar7
  include: /longpath/linux-4.4/arch/mips/include/uapi
  include: /longpath/linux-4.4/arch/mips/include/asm/mach-generic
  include: /longpath/linux-4.4/arch/mips/include/asm
  include: /longpath/linux-4.4/include/linux
  define: CONFIG_BASE_BODEGA = 1
  define: CONFIG_BASE = BODEGA
  define: config_base = bodega
  other: --gcc
  other: --gnu_version=90100
  stdver: c11
  intelliSenseMode: gcc-x86
Checking for syntax errors: file:///longpath/linux-4.4/fs/inode.c
Queueing IntelliSense update for files in translation unit of: /longpath/linux-4.4/fs/inode.c
idle loop: reparsing the active document
Checking for syntax errors: file:///longpath/linux-4.4/fs/inode.c
Queueing IntelliSense update for files in translation unit of: /longpath/linux-4.4/fs/inode.c
cpptools/finishUpdateSquiggles
Error squiggle count: 449
Error squiggles will be disabled in: file:///longpath/linux-4.4/fs/inode.c
terminating child process: 631
Update IntelliSense time (sec): 1.96
cpptools/getCodeActions: file:///longpath/linux-4.4/fs/inode.c (id: 9)

**Comment by sean-mcmanus** [comment]
The "Compiler does not support 32-bit." message is incorrect/misleading and in this case just means the compiler doesn't support the -m32 flag. It looks like gcc-x86 is being used (correctly).

I'll create a new issue tracking better error reporting when an invalid compilerArg is used: https://github.com/microsoft/vscode-cpptools/issues/5893

**Comment by UberUser** [comment]
Thanks for seeing this through.

I was able to fix the 32-bit error by switching the compiler type to gcc-arm.

While investigating I noticed that gcc supports at least two dozen architectures. Obviously it'd be impractical to support more than a few, which I think argues for generic 32 and 64 bit options. Then every possible architecture can be "supported", with a few at a higher level of support than most.
